Automattic/jetpack
# Boost: Add more context to api fetch error

Fixes #20946

#### Changes proposed in this Pull Request:

The following error was reported 2 times on the Jetpack Boost support forum ((here and here)) when a user tried to generate Critical CSS.

```
Failed to generate Critical CSS
Hmm, looks like something went wrong. We’re still seeing an unexpected error. Please reach out to our support to get help.

Hide error message
TypeError: Failed to fetch
```

Unfortunately, this error is quite obscure and difficult to know the cause. 

Originally I thought that this error was caused by a fetch call in the code from the [jetpack-boost-critical-css-gen](https://github.com/Automattic/jetpack-boost-critical-css-gen) repo [here](https://github.com/Automattic/jetpack-boost-critical-css-gen/blob/trunk/lib/browser-interface-iframe.js#L87) or [here](https://github.com/Automattic/jetpack-boost-critical-css-gen/blob/trunk/lib/css-file-set.js#L57) but after some tests, it seems that as these fetch calls surrounded by a try/catch statement and the error repurposed, then the error cannot come from here.

The other place where the error may come from is the [fetch call in the api.ts](https://github.com/Automattic/jetpack/blob/master/projects/plugins/boost/app/assets/src/js/api/api.ts#L34) file from the Jetpack Boost plugin. Possible errors with this call are caught at a higher level and only the error itself is captured. As a result, if it failed as reported we only get the `TypeError: Failed to fetch` message displayed to the user for reporting but it is not really useful. 

With the PR I am attempting to catch the error when the call is done, create a more useful error with extra context to it so it can be displayed to the user and provide more information to us when reported.

![image](https://user-images.githubusercontent.com/927932/131958157-6a84dbee-c13d-4084-ac3d-fbc754b65139.png)


#### Jetpack product discussion


N/A

#### Does this pull request change what data or activity we track or use?


N/A

#### Testing instructions:






It is difficult to test because this error as I actually could not work out the cause for it. But there is a way to 'kind of' simulate it.

* In this PR change [line 39 of the api.ts file](https://github.com/Automattic/jetpack/blob/improve/add-more-context-to-api-fetch-error/projects/plugins/boost/app/assets/src/js/api/api.ts#L39) from
```
const endpointFullUrl = getEndpointUrl( path );
```

to 
```
let endpointFullUrl = getEndpointUrl( path );
	if ( path.includes( 'success' ) ) {
		endpointFullUrl = endpointFullUrl.replace( '://', '://admin.' );
	}
```

* Regenerate Critical Css 
* See a more detailed error message being displayed than just `TypeError: Failed to fetch`
